File: src/main/java/fortress/tfol/operations/TypeChecker.java
package fortress.tfol.operations;

import java.util.Optional;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.Set;
import java.util.HashSet;
import fortress.tfol.*;
import java.util.Iterator;
import fortress.util.Pair;
import fortress.data.ImmutableList;
import fortress.data.TypeCheckException;

public class TypeChecker {
    
    public TypeChecker() { }
    
    /**
     * Given a signature and a term, typechecks the term with respect to the signature.
     * Returns a TypeCheckResult containing the type of the term, AND a new term
     * that is equal to the old term but with instances of Eq replaced with Iff
     * when comparing Bool types. Such a term is called "sanitized".
    */
    public static TypeCheckResult typeCheck(Signature signature, Term term) {
        return new TypeCheckVisitor(signature).visit(term);
    }
    
    // Visitor interface hidden -- the fact that it uses a visitor is
    // just an implementation detail
    private static class TypeCheckVisitor extends TermVisitorWithTypeContext<TypeCheckResult> {
        
        public TypeCheckVisitor(Signature signature) {
            super(signature);
        }
    
        @Override
        public TypeCheckResult visitTop(Top top) {
            return new TypeCheckResult(top, Type.Bool,
                /* containsConnectives */ false, /* containsQuantifiers */ false);
        }
        
        @Override
        public TypeCheckResult visitBottom(Bottom bottom) {
            return new TypeCheckResult(bottom, Type.Bool,
                /* containsConnectives */ false, /* containsQuantifiers */ false);
        }
        
        @Override
        public TypeCheckResult visitVar(Var variable) {
            // Check variable is not an already declared function symbol
            // This must be done even with a consistent signature
            // TODO: this behaviour should be documented
            // TODO: is this considered poorly typed or a different kind of error?
            if(lookupFunctionDeclaration(variable.getName()).isPresent()) {
                throw new TypeCheckException.NameConflict("Variable or constant name " + variable.getName() + " conflicts with existing function symbol");
            }
            
            // Check variable is not an already declared type symbol
            if(signature.hasType(variable.getName())) {
                throw new TypeCheckException.NameConflict("Variable or constant name " + variable.getName() + " conflicts with existing type symbol");
            }
            
            Optional<Type> typeMaybe = lookupType(variable);
            if(typeMaybe.isPresent()) {
                return new TypeCheckResult(variable, typeMaybe.get(),
                    /* containsConnectives */ false, /* containsQuantifiers */ false);
            } else {
                throw new TypeCheckException.UndeterminedType("Could not determine type of variable " + variable.getName());
            }
        }
        
        @Override
        public TypeCheckResult visitNot(Not not) {
            TypeCheckResult bodyResult = visit(not.getBody());
            Term newBody = bodyResult.term;
            Type bodyType = bodyResult.type;
            if(bodyType.equals(Type.Bool)) {
                return new TypeCheckResult(Term.mkNot(newBody), Type.Bool,
                    /* containsConnectives */ true, /* containsQuantifiers */ bodyResult.containsQuantifiers);
            } else {
                throw new TypeCheckException.WrongArgType("Argument of negation is of type " + bodyType.getName() + " in " + not.toString());
            }
        }

        @Override
        public TypeCheckResult visitAndList(AndList andList) {
            ImmutableList<TypeCheckResult> results = andList.getArguments().map(this::visit);
            for(TypeCheckResult r : results) {
                if(! r.type.equals(Type.Bool)) {
                    throw new TypeCheckException.WrongArgType("Expected type Bool but was " + r.type.getName() + " in " + andList.toString());
                }
            }
            ImmutableList<Term> newArguments = results.map(r -> r.term);
            boolean containsQuantifiers = results.stream().anyMatch((TypeCheckResult r) -> r.containsQuantifiers);
            return new TypeCheckResult(Term.mkAndF(newArguments), Type.Bool,
                /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
        }
        
        @Override
        public TypeCheckResult visitOrList(OrList orList) {
            ImmutableList<TypeCheckResult> results = orList.getArguments().map(this::visit);
            for(TypeCheckResult r : results) {
                if(! r.type.equals(Type.Bool)) {
                    throw new TypeCheckException.WrongArgType("Expected type Bool but was " + r.type.getName() + " in " + orList.toString());
                }
            }
            ImmutableList<Term> newArguments = results.map(r -> r.term);
            boolean containsQuantifiers = results.stream().anyMatch((TypeCheckResult r) -> r.containsQuantifiers);
            return new TypeCheckResult(Term.mkOrF(newArguments), Type.Bool,
                /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
        }
        
        @Override
        public TypeCheckResult visitDistinct(Distinct distinct) {
            Set<Type> foundTypes = new HashSet();
            ImmutableList<TypeCheckResult> results = distinct.getArguments().map(this::visit);
            
            for(TypeCheckResult result : results) {
                foundTypes.add(result.type);
            }
            
            if(foundTypes.size() > 1) {
                throw new TypeCheckException.WrongArgType("Arguments of multiple types " + foundTypes.toString() + " in " + distinct.toString());
            }
            
            ImmutableList<Term> newArguments = results.map(r -> r.term);
            boolean containsQuantifiers = results.stream().anyMatch((TypeCheckResult r) -> r.containsQuantifiers);
            return new TypeCheckResult(Term.mkDistinctF(newArguments), Type.Bool,
                /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
        }
        
        @Override
        public TypeCheckResult visitImplication(Implication imp) {
            TypeCheckResult leftResult = visit(imp.getLeft());
            TypeCheckResult rightResult = visit(imp.getRight());
            if(! leftResult.type.equals(Type.Bool)) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + leftResult.type.getName() + " in " + imp.toString() );
            }
            if(! rightResult.type.equals(Type.Bool)) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + rightResult.type.getName() + " in " + imp.toString() );
            }
            boolean containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers;
            return new TypeCheckResult(Term.mkImp(leftResult.term, rightResult.term), Type.Bool,
                /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
        }
        @Override
        public TypeCheckResult visitIff(Iff iff) {
            TypeCheckResult leftResult = visit(iff.getLeft());
            TypeCheckResult rightResult = visit(iff.getRight());
            if(! leftResult.type.equals(Type.Bool)) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + leftResult.type.getName() + " in " + iff.toString() );
            }
            if(! rightResult.type.equals(Type.Bool)) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + rightResult.type.getName() + " in " + iff.toString() );
            }
            boolean containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers;
            return new TypeCheckResult(Term.mkIff(leftResult.term, rightResult.term), Type.Bool,
                /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
        }
        
        @Override
        public TypeCheckResult visitEq(Eq eq) {
            Set<Type> foundTypes = new HashSet();
            TypeCheckResult leftResult = visit(eq.getLeft());
            TypeCheckResult rightResult = visit(eq.getRight());
            
            if(! leftResult.type.equals(rightResult.type)) {
                throw new TypeCheckException.WrongArgType("Mismatched argument types " + leftResult.type.toString() + " and "
                    + rightResult.type.toString() + " in " + eq.toString());
            }
            boolean containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers;
            // Replaces (Bool) = (Bool) with Iff
            if(leftResult.type.equals(Type.Bool)) {
                return new TypeCheckResult(Term.mkIff(leftResult.term, rightResult.term), Type.Bool,
                    /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
            } else {
                return new TypeCheckResult(Term.mkEq(leftResult.term, rightResult.term), Type.Bool,
                    /* containsConnectives */ true, /* containsQuantifiers*/ containsQuantifiers);
            }
        }
        
        // Check argument:
        // 1. types match function declaration
        // 2. arguments contain no connectives or quantifiers
        private TypeCheckResult checkFunction(FuncDecl funcDecl, ImmutableList<Term> arguments) {
            if(funcDecl.getArity() != arguments.size()) {
                throw new TypeCheckException.WrongArity("Application of " + funcDecl.getName() + " to wrong number of arguments");
            }
            
            ImmutableList<TypeCheckResult> results = arguments.map(this::visit);
            
            Iterator<Type> itTypes = funcDecl.getArgTypes().iterator();
            Iterator<TypeCheckResult> itResults = results.iterator();
            int argNum = 0;
            while(itTypes.hasNext() && itResults.hasNext()) {
                argNum++;
                Type expected = itTypes.next();
                TypeCheckResult result = itResults.next();
                Type actual = result.type;
                if(!expected.equals(actual)) {
                    throw new TypeCheckException.WrongArgType("In argument " + argNum + " of " + funcDecl.getName() + ", expected " + expected.getName() + " but was " + actual.getName());
                }
                
                if(result.containsConnectives) {
                    throw new TypeCheckException.BadStructure("Argument " + argNum + " of " + funcDecl.getName() + " contains connective");
                }
                
                if(result.containsQuantifiers) {
                    throw new TypeCheckException.BadStructure("Argument " + argNum + " of " + funcDecl.getName() + " contains quantifier");
                }
            }
            
            ImmutableList<Term> newArguments = results.map(r -> r.term);
            return new TypeCheckResult(Term.mkAppF(funcDecl.getName(), newArguments), funcDecl.getResultType(),
                /* containsConnectives */ false, /* containsQuantifiers */ false);
        }
        
        @Override
        public TypeCheckResult visitApp(App app) {
            String funcName = app.getFunctionName();
            
            // Check for function symbol in declared functions
            Optional<FuncDecl> fMaybe = lookupFunctionDeclaration(funcName);
            if(! fMaybe.isPresent()) {
                throw new TypeCheckException.UnknownFunction("Unknown function " + funcName);
            } else {
                FuncDecl fdecl = fMaybe.get();
                return checkFunction(fdecl, app.getArguments());
            }
        }
        
        @Override
        public TypeCheckResult visitExistsInner(Exists exists) {
            // Check variables don't clash with function names
            // and that their type exists
            for(AnnotatedVar av : exists.getVars()) {
                if(lookupFunctionDeclaration(av.getName()).isPresent()) {
                    throw new TypeCheckException.NameConflict("Variable name " + av.getName() + " conflicts with existing function symbol");
                }
                if(! signature.hasType(av.getType())) {
                    throw new TypeCheckException.UnknownType("Unknown type " + av.getType().getName() + " in " + exists.toString());
                }
            }
            
            TypeCheckResult bodyResult = visit(exists.getBody());
            if(bodyResult.type.equals(Type.Bool)) {
                return new TypeCheckResult(Term.mkExists(exists.getVars(), bodyResult.term), Type.Bool,
                    /* containsConnective */ bodyResult.containsConnectives, /* containsQuantifiers */ true);
            } else {
                throw new TypeCheckException.WrongArgType("Expected Bool but was " + bodyResult.type.toString() + " in " + exists.toString());
            }
        }
        
        @Override
        public TypeCheckResult visitForallInner(Forall forall) {
            // Check variables don't clash with function names
            // and that their type exists
            for(AnnotatedVar av : forall.getVars()) {
                if(lookupFunctionDeclaration(av.getName()).isPresent()) {
                    throw new TypeCheckException.NameConflict("Variable name " + av.getName() + " conflicts with existing function symbol");
                }
                if(! signature.hasType(av.getType())) {
                    throw new TypeCheckException.UnknownType("Unknown type " + av.getType().getName() + " in " + forall.toString());
                }
            }
            
            TypeCheckResult bodyResult = visit(forall.getBody());
            if(bodyResult.type.equals(Type.Bool)) {
                return new TypeCheckResult(Term.mkForall(forall.getVars(), bodyResult.term), Type.Bool,
                    /* containsConnective */ bodyResult.containsConnectives, /* containsQuantifiers */ true);
            } else {
                throw new TypeCheckException.WrongArgType("Expected Bool but was " + bodyResult.type.toString() + " in " + forall.toString());
            }
        }
    
    }
    
}

File: src/main/scala/fortress/tfol/operations/TypeChecker.scala
package fortress.tfol.operations

import scala.collection.immutable.Seq // Use immutable seq by default

import fortress.util.Errors
import fortress.tfol._
import fortress.data._

case class TypeCheckResult(
    sanitizedTerm: Term,
    sort: Type,
    containsConnectives: Boolean,
    containsQuantifiers: Boolean
)

/** Given a signature and a term, typechecks the term with respect to the signature.
 * Returns a TypeCheckResult containing the type of the term, AND a new term
 * that is equal to the old term but with instances of Eq replaced with Iff
 * when comparing Bool types. Such a term is called "sanitized". */
class TypeChecker(signature: Signature) extends TermVisitorWithTypeContext[TypeCheckResult](signature) {
    override def visitTop: TypeCheckResult =
        TypeCheckResult(sanitizedTerm = Top, sort = BoolType, containsConnectives = false, containsQuantifiers = false)
        
    override def visitBottom: TypeCheckResult =
        TypeCheckResult(sanitizedTerm = Bottom, sort = BoolType, containsConnectives = false, containsQuantifiers = false)
    
    override def visitVar(variable: Var): TypeCheckResult = {
        // Check variable is not an already declared function symbol
        // This must be done even with a consistent signature
        // TODO: this behaviour should be documented
        // TODO: is this considered poorly typed or a different kind of error?
        if(signature hasFunctionWithName variable.name) {
            throw new TypeCheckException.NameConflict("Variable or constant name " + variable.name + " conflicts with existing function symbol")
        }
        
        // Check variable is not an already declared type symbol
        if(signature hasTypeWithName variable.name) {
            throw new TypeCheckException.NameConflict("Variable or constant name " + variable.name + " conflicts with existing type symbol")
        }
        
        val typeMaybe = lookupType(variable)
        if(!typeMaybe.isPresent()) {
            throw new TypeCheckException.UndeterminedType("Could not determine type of variable " + variable.name)
        }
        
        TypeCheckResult(sanitizedTerm = variable, sort = typeMaybe.get,
            containsConnectives = false, containsQuantifiers = false)
    }
    
    override def visitNot(not: Not): TypeCheckResult = {
        val bodyResult = visit(not.body)
        if(bodyResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Argument of negation is of type " + bodyResult.sort.name + " in " + not.toString)
        }
        TypeCheckResult(sanitizedTerm = Not(bodyResult.sanitizedTerm), sort = BoolType,
            containsConnectives = true, containsQuantifiers = bodyResult.containsQuantifiers)
    }
    
    override def visitAndList(andList: AndList): TypeCheckResult = {
        val results = andList.arguments.map(visit)
        for(result <- results) {
            if(result.sort != BoolType) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + result.sort.name + " in " + andList.toString)
            }
        }
        TypeCheckResult(sanitizedTerm = AndList(results.map(_.sanitizedTerm)), sort = BoolType,
            containsConnectives = true,
            containsQuantifiers = results.exists(_.containsQuantifiers))
    }
    
    override def visitOrList(orList: OrList): TypeCheckResult = {
        val results = orList.arguments.map(visit)
        for(result <- results) {
            if(result.sort != BoolType) {
                throw new TypeCheckException.WrongArgType("Expected type Bool but was " + result.sort.name + " in " + orList.toString)
            }
        }
        TypeCheckResult(sanitizedTerm = OrList(results.map(_.sanitizedTerm)), sort = BoolType,
            containsConnectives = true,
            containsQuantifiers = results.exists(_.containsQuantifiers))
    }
    
    override def visitDistinct(distinct: Distinct): TypeCheckResult = {
        val results: Seq[TypeCheckResult] = distinct.arguments.map(visit)
        
        // Check sorts all the same
        val sorts: Seq[Type] = results.map(_.sort)
        if(sorts.distinct.size > 1) {
            throw new TypeCheckException.WrongArgType("Arguments of multiple types " + sorts.toString + " in " + distinct.toString)
        }
        
        TypeCheckResult(sanitizedTerm = Distinct(results.map(_.sanitizedTerm)), sort = BoolType,
            containsConnectives = true,
            containsQuantifiers = results.exists(_.containsQuantifiers))
    }
    
    override def visitImplication(imp: Implication): TypeCheckResult = {
        val leftResult = visit(imp.left)
        val rightResult = visit(imp.right)
        
        if(leftResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected type Bool but was " + leftResult.sort.name + " in " + imp.toString)
        }
        if(rightResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected type Bool but was " + rightResult.sort.name + " in " + imp.toString)
        }
        TypeCheckResult(sanitizedTerm = Implication(leftResult.sanitizedTerm, rightResult.sanitizedTerm), sort = BoolType,
            containsConnectives = true, containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers)
    }
    
    override def visitIff(iff: Iff): TypeCheckResult = {
        val leftResult = visit(iff.left)
        val rightResult = visit(iff.right)
        
        if(leftResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected type Bool but was " + leftResult.sort.name + " in " + iff.toString)
        }
        if(rightResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected type Bool but was " + rightResult.sort.name + " in " + iff.toString)
        }
        TypeCheckResult(sanitizedTerm = Iff(leftResult.sanitizedTerm, rightResult.sanitizedTerm), sort = BoolType,
            containsConnectives = true, containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers)
    }
    
    override def visitEq(eq: Eq): TypeCheckResult = {
        val leftResult = visit(eq.left)
        val rightResult = visit(eq.right)
        
        if(leftResult.sort != rightResult.sort) {
            throw new TypeCheckException.WrongArgType("Mismatched argument types " + leftResult.sort.toString + " and "
                + rightResult.sort.toString + " in " + eq.toString)
        }
        
        // Replaces (Bool) = (Bool) with Iff
        val sanTerm =
            if (leftResult.sort == BoolType) Iff(leftResult.sanitizedTerm, rightResult.sanitizedTerm)
            else Eq(leftResult.sanitizedTerm, rightResult.sanitizedTerm)
        
        TypeCheckResult(sanitizedTerm = sanTerm, sort = BoolType,
            containsConnectives = true, containsQuantifiers = leftResult.containsQuantifiers || rightResult.containsQuantifiers)
    }
    
    override def visitApp(app: App): TypeCheckResult = {
        // Check argument:
        // 1. types match function declaration
        // 2. arguments contain no connectives or quantifiers
        val funcName = app.functionName
        
        if(! (signature hasFunctionWithName funcName) ) {
            throw new TypeCheckException.UnknownFunction("Could not find function " + funcName)
        }
        
        val results = app.arguments.map(visit)
        val argSorts = results.map(_.sort)
        val fdecl = signature.queryFunction(funcName, argSorts) match {
            case None => throw throw new TypeCheckException.WrongArgType(funcName + " cannot accept argument types " + argSorts.toString + " in " + app.toString)
            case Some(fdecl) => fdecl
        }
        
        if(results.exists(_.containsConnectives)) {
            throw new TypeCheckException.BadStructure("Argument of " + funcName + " contains connective")
        }
        if(results.exists(_.containsQuantifiers)) {
            throw new TypeCheckException.BadStructure("Argument of " + funcName + " contains quantifier")
        }
        
        TypeCheckResult(sanitizedTerm = App(funcName, results.map(_.sanitizedTerm)), sort = fdecl.resultType,
            containsConnectives = false, containsQuantifiers = false)
    }
    
    override def visitExistsInner(exists: Exists): TypeCheckResult = {
        // Check variables don't clash with function names
        // and that their type exists
        for(av <- exists.vars) {
            if(signature hasFunctionWithName av.name) {
                throw new TypeCheckException.NameConflict("Variable name " + av.name + " conflicts with existing function symbol")
            }
            if(! (signature hasType av.sort) ) {
                throw new TypeCheckException.UnknownType("Unknown type " + av.sort.name + " in " + exists.toString)
            }
        }
        val bodyResult = visit(exists.body)
        if(bodyResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected Bool but was " + bodyResult.sort.name + " in " + exists.toString)
        }
        TypeCheckResult(sanitizedTerm = Exists(exists.vars, bodyResult.sanitizedTerm), sort = BoolType,
            containsConnectives = bodyResult.containsConnectives, containsQuantifiers = true)
    }
    
    override def visitForallInner(forall: Forall): TypeCheckResult = {
        // Check variables don't clash with function names
        // and that their type exists
        for(av <- forall.vars) {
            if(signature hasFunctionWithName av.name) {
                throw new TypeCheckException.NameConflict("Variable name " + av.name + " conflicts with existing function symbol")
            }
            if(! (signature hasType av.sort) ) {
                throw new TypeCheckException.UnknownType("Unknown type " + av.sort.name + " in " + forall.toString)
            }
        }
        val bodyResult = visit(forall.body)
        if(bodyResult.sort != BoolType) {
            throw new TypeCheckException.WrongArgType("Expected Bool but was " + bodyResult.sort.name + " in " + forall.toString)
        }
        TypeCheckResult(sanitizedTerm = Forall(forall.vars, bodyResult.sanitizedTerm), sort = BoolType,
            containsConnectives = bodyResult.containsConnectives, containsQuantifiers = true)
    }
    
    override def visitDomainElement(d: DomainElement): TypeCheckResult = {
        if(! (signature hasType d.sort)) {
            throw new TypeCheckException.UnknownType("Unkown type " + d.sort.name + " in " + d.toString)
        }
        TypeCheckResult(sanitizedTerm = d, sort = d.sort, containsConnectives = false, containsQuantifiers = false)
    }
    
    override def visitIntegerLiteral(literal: IntegerLiteral): TypeCheckResult = {
        // Should fail if Int is not in signature
        if(! (signature hasType IntType) ) {
            throw new TypeCheckException.UnknownType("Given signature does not include integers")
        } else {
           TypeCheckResult(sanitizedTerm = literal, sort = IntType,
                containsConnectives = false, containsQuantifiers = false)
        }
    }
    
    override def visitBitVectorLiteral(literal: BitVectorLiteral): TypeCheckResult = {
        // Should fail if BitVector is not in signature
        if(! (signature hasType BitVectorType(literal.bitWidth)) ) {
            throw new TypeCheckException.UnknownType("Given signature does not include bit vectors")
        } else {
           TypeCheckResult(sanitizedTerm = literal, sort = BitVectorType(literal.bitWidth),
                containsConnectives = false, containsQuantifiers = false)
        }
    }
    
    override def visitEnumValue(e: EnumValue) = signature.queryEnum(e) match {
        case Some(eSort: Type) => TypeCheckResult(sanitizedTerm = e, sort = eSort, containsConnectives = false, containsQuantifiers = true)
        case None => throw new TypeCheckException.UndeterminedType("Could not determine type of enum " + e.name)
    }
    
}
